hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -179,6 +194,9 @@ object HoodieSparkSqlWriter {
           basePath, path, instantTime, partitionColumns)
         return (success, commitTime, common.util.Option.empty(), common.util.Option.empty(), hoodieWriteClient.orNull, tableConfig)
       }
+      if (tableExists && tableConfig.isAppendOnlyTable && operation != WriteOperationType.BULK_INSERT) {
```

Comment:
Conditional above is short-circuiting for bulk-insert with row-writing.

We should do this check as early as possible. We can also extract it to a `validateConfig` method